devito/builtins.py

```diff
@@ -53,6 +54,135 @@ def smooth(f, g, axis=None):
+        slices = tuple([slice(lw, -lw) for _ in range(fi.grid.dim)])
+        fi.data[slices] = data
+        eqs = []
+        for d in fi.dimensions:
+            dim_l = dv.SubDimension.left(name='abc_%s_l' % d.name, parent=d,
+                                         thickness=lw)
+            dim_r = dv.SubDimension.right(name='abc_%s_r' % d.name, parent=d,
+                                          thickness=lw)
+            if mode == 'constant':
+                to_copy = lw
+                eqs += [dv.Eq(fi.subs({d: dim_l}), fi.subs({d: to_copy}))]
+                to_copy = d.symbolic_max - lw
+                eqs += [dv.Eq(fi.subs({d: dim_r}), fi.subs({d: to_copy}))]
+                rhs = dv.generic_derivative(fi, d, 2*lw, 1)
+                coeffs = dv.Coefficient(1, fi, d, weights)
+                eqs += [dv.Eq(fo, rhs, coefficients=dv.Substitutions(coeffs),
+                              subdomain=grid.subdomains['objective_domain'])]
+            elif mode == 'reflect':
+                to_copy = 2*lw - 1 - dim_l
+                eqs += [dv.Eq(fi.subs({d: dim_l}), fi.subs({d: to_copy}))]
+                to_copy = 2*(d.symbolic_max - lw) + 1 - dim_r
+                eqs += [dv.Eq(fi.subs({d: dim_r}), fi.subs({d: to_copy}))]
+                rhs = dv.generic_derivative(fi, d, 2*lw, 1)
+                coeffs = dv.Coefficient(1, fi, d, weights)
+                eqs += [dv.Eq(fo, rhs, coefficients=dv.Substitutions(coeffs),
+                              subdomain=grid.subdomains['objective_domain'])]
+            else:
+                raise ValueError("Mode not available")
+            eqs += [dv.Eq(fi, fo, subdomain=grid.subdomains['objective_domain'])]
+
+        dv.Operator(eqs)()
+
+    def fset(f, g):
+        indices = [slice(lw, -lw, 1) for _ in g.grid.dimensions]
+        slices = (slice(None, None, 1), )*len(g.grid.dimensions)
+        if isinstance(f, np.ndarray):
+            f[slices] = g.data[tuple(indices)]
+        elif isinstance(f, dv.Function):
+            f.data[slices] = g.data[tuple(indices)]
+        else:
+            raise NotImplementedError
+
+    lw = int(_order*sigma + 0.5)
+
+    # Create the padded grid:
+    objective_domain = ObjectiveDomain(lw)
+    try:
+        shape_padded = np.array(f.grid.shape) + 2*lw
+    except AttributeError:
+        shape_padded = np.array(f.shape) + 2*lw
+    grid = dv.Grid(shape=shape_padded, subdomains=objective_domain)
+
+    f_c = dv.Function(name='f_c', grid=grid, space_order=2*lw,
+                      coefficients='symbolic', dtype=np.int32)
+    f_o = dv.Function(name='f_o', grid=grid, coefficients='symbolic', dtype=np.int32)
+
+    weights = np.exp(-0.5/sigma**2*(np.linspace(-lw, lw, 2*lw+1))**2)
+    weights = weights/weights.sum()
+
+    if isinstance(f, dv.Function):
+        smooth(f_o, f_c, f.data[:])
+    else:
+        smooth(f_o, f_c, f)
+
+    fset(f, f_o)
+
+    return f
+
+
+def initialize_function(function, data, nbpml, mode='constant'):
```

Comment:
Maybe this should be squashed with the `assign` builtin so as to create something unique and sufficiently general for all our needs ? otherwise, there's an overlap between these builtins...although their names are totally different.

Reply:
Yeah, should probably do something along those lines.